src/System.Private.ServiceModel/tests/Common/Infrastructure/src/ConditionalWcfTest.cs

```diff
@@ -10,42 +15,191 @@ namespace Infrastructure.Common
+        private static ConcurrentDictionary<string, bool> _evaluatedConditions = new ConcurrentDictionary<string, bool>();
+
+        // Returns 'true' or 'false' for the given condition name.
+        // If this is the first call and 'detectFunc' is specified, it will be
+        // called to detect whether the condition is true or false.  If not
+        // specified, it will default to false. The value is cached after first call.
+        private static bool GetConditionValue(string conditionName, Func<bool> detectFunc = null)
         {
-            string propertyValue = TestProperties.GetProperty(testCategoryName);
-            bool isEnabled;
-            if (!String.IsNullOrEmpty(propertyValue) && bool.TryParse(propertyValue, out isEnabled))
+            return _evaluatedConditions.GetOrAdd(conditionName, (n) =>
             {
-                return isEnabled;
-            }
+                // We allow all conditions to be satisfied through an environment variable
+                // if it has been set to 'true' or 'false'.  If not set, fall into detection logic.
+                // This convention permits a developer to override any detection logic to
+                // enable or disable manually.
+                string environmentVariable = Environment.GetEnvironmentVariable(conditionName);
+                bool result = false;
+                if (!String.IsNullOrWhiteSpace(environmentVariable) && bool.TryParse(environmentVariable, out result))
+                {
+                    return result;
+                }
+                
+                // No enviromnent variable, so detect at runtime or default to 'false'
+                return detectFunc == null ? false : detectFunc();
+            });
+        }
 
-            return false;
+        // Returns the 'true' or 'false' for a condition whose name is
+        // contained in TestProperties.   If the property's value is
+        // 'true' or 'false', it will be used as is.  If the value is
+        // empty, and 'detectFunc' is specified, it will be invoked to
+        // evaluate the condition at runtime.  The value is cached, so
+        // 'detectFunc' will be called only once.
+        private static bool GetConditionValueFromTestProperties(string conditionName, Func<bool> detectFunc = null)
+        {
+            string propertyValue;
+            return GetConditionValue(conditionName, () =>
+            {
+                propertyValue = TestProperties.GetProperty(conditionName);
+                if (string.IsNullOrWhiteSpace(propertyValue))
+                {
+                    // Empty string means auto-detect if possible, else false
+                    return (detectFunc == null) ? false : detectFunc();
+                }
+                else
+                {
+                    bool conditionResult = false;
+                    if (!bool.TryParse(propertyValue, out conditionResult))
```

Comment:
@iamjasonp I have a different opinion. If a condition is specified explicitly through either TestProperty or environment variable but the value is invalid, we should always use false instead of trying to be smart (auto detection). This will make sure we have consistent results and also help to discover the invalid value (most likely unintentionally).

I think the problem here is we don't have an agreement on design yet. I suggest us to conclude [this design discussion](https://github.com/dotnet/wcf/pull/1026#issuecomment-211263175) before proceeding.